File: leetcode/cpp/386.lexicographical-numbers.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> ans;
        ans.reserve(n+1);
        for(int i = 1; i < 10 && i <= n; i++ ) {
            genNums(i, n, ans);
        }
        return ans;
    }
private:
    void genNums(int x, int n, vector<int>& ans) {
        ans.push_back(x);
        for(int y=x * 10; y < x * 10 + 10 && y <= n; y++) {
            genNums(y, n, ans);
        }
    }
};

File: leetcode/python/386.lexicographical-numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        ans = []
        def genNums(i: int):
            ans.append(i)
            for j in range(i*10, i*10+10):
                if j > n:
                    break
                genNums(j)
        for x in range(1, 10):
            if x > n:
                break
            genNums(x)
        return ans

